### PR TITLE
Add JunoVinylService:

### PR DESCRIPTION
**Pull request by TEdwardJ**
- Add a new vinyl service implementation for juno.co.uk, add it to shops parser;
 - Add Juno to the shops list;

### CODE REVIEW

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    Pattern pageNumberPattern = Pattern.compile("/([0-9]+)/");
+
+    @Override
+    public List<RawOffer> getRawOffersList() {
+        Set<String> pageLinks = getPresentPageLinks();
+
+        Set<RawOffer> RawOffers = readVinylsDataFromAllPages(pageLinks);
+        List<RawOffer> RawOffersJunoRawOfferList = new ArrayList<>(RawOffers);
+        log.debug("Resulting list of RawOffers from RawOffer.ua is {'RawOffersRawOfferUaList':{}}", RawOffersJunoRawOfferList);
```
Comment (Greengrer) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: Resulting list of rawOffers from juno.co.uk is {'rawOffers':{}}

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    private final String RawOfferInfoItemsSelector = "div.vi-text";
+
+    Pattern pageNumberPattern = Pattern.compile("/([0-9]+)/");
+
+    @Override
+    public List<RawOffer> getRawOffersList() {
+        Set<String> pageLinks = getPresentPageLinks();
+
+        Set<RawOffer> RawOffers = readVinylsDataFromAllPages(pageLinks);
+        List<RawOffer> RawOffersJunoRawOfferList = new ArrayList<>(RawOffers);
```
Comment (Greengrer) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: junoRawOffers/rawOffersList/junoRawOffersList

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        log.debug("Resulting list of RawOffers from RawOffer.ua is {'RawOffersRawOfferUaList':{}}", RawOffersJunoRawOfferList);
+        return RawOffersJunoRawOfferList;
```
Comment (Greengrer) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: junoRawOffers/rawOffersList/junoRawOffersList

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            var RawOfferLink = item.select("a").get(0).attr("href");
+            var info = item.select(RawOfferInfoBlockSelector);
+            var infoDetails = info.select(RawOfferInfoItemsSelector);
+            var priceBlock = item.select(priceBlockSelector);
+            var priceString = priceBlock.select(priceLineSelector).text();
+            var price = extractPrice(priceString);
+            var priceCurrency = "GBP";
+            var artist = infoDetails.get(0).select("a").text();
+            var release = infoDetails.get(1).select("a").text();
+            var genre = infoDetails.get(4).text();
```
Comment (Greengrer) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java:102]: @AlexKievDeveloper should we use String, or is var sufficient?

**Review by Greengrer — CHANGES_REQUESTED** [review]
Okay, that is pretty cool.

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    private final String RawOfferItemsListSelector = "div.product-list";
+    private final String priceBlockSelector = "div.pl-buy";
+    private final String priceLineSelector = "span.price_lrg.text-cta";
+    private final String RawOfferItemSelector = "div.dv-item";
+    private final String imageLinkSelector = "img.lazy_img.img-fluid";
+    private final String RawOfferInfoBlockSelector = "div.pl-info";
+    private final String RawOfferInfoItemsSelector = "div.vi-text";
+
+    Pattern pageNumberPattern = Pattern.compile("/([0-9]+)/");
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: Can we add private access modifier?

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+                .map(this::getDocument)
+                .filter(Objects::nonNull)
+                .map(document -> document.select(RawOfferItemsListSelector))
+                .flatMap(RawOfferItemsList -> RawOfferItemsList.select(RawOfferItemSelector).stream())
+                .map(this::itemToRawOffer)
+                .filter(Objects::nonNull)
+                .collect(toSet());
+        log.debug("Resulting set of rawOffer  is {'rawOfferSet':{}}", rawOfferSet);
+        return rawOfferSet;
+
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: Please remove empty row

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        var tmpPrice = priceString.replaceAll("^[^0-9]+", "");
+        return Double.parseDouble(tmpPrice);
+    }
+
+    private Document getDocument(String url) {
+        try {
+            return Jsoup.connect(url).get();
+        } catch (IOException e) {
+            log.error("Error while getting document by link {'link':{}}", url, e);
+            return null;
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: We shouldn`t return null. Let`s use optional

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            rawOffer.setArtist(artist);
+            rawOffer.setPrice(price);
+            rawOffer.setCurrency(Currency.getCurrency(priceCurrency));
+            rawOffer.setOfferLink(RawOfferLink);
+            rawOffer.setImageLink(imageLink);
+            rawOffer.setGenre(genre);
+            return rawOffer;
+        } catch(Exception e){
+            log.error("Error during RawOffer creation from the HTML element {}", item, e);
+            return null;
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: We shouldn`t return null. Better optional

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+            rawOffer.setRelease(release);
+            rawOffer.setArtist(artist);
+            rawOffer.setPrice(price);
+            rawOffer.setCurrency(Currency.getCurrency(priceCurrency));
+            rawOffer.setOfferLink(RawOfferLink);
+            rawOffer.setImageLink(imageLink);
+            rawOffer.setGenre(genre);
+            return rawOffer;
+        } catch(Exception e){
+            log.error("Error during RawOffer creation from the HTML element {}", item, e);
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: Don`t we need test on this method and other private methods?

**Review by AlexKievDeveloper — CHANGES_REQUESTED** [review]

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java (hunk 1 above)
Comment (AlexKievDeveloper) [reply src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: > Resulting list of rawOffers from juno.co.uk is {'rawOffers':{}}

molodec! vnimatelnaya! (c) Tolik

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    private final String startBaseLink = "https://www.juno.co.uk/all/";
+    private final String startLink = startBaseLink + "back-cat/2/?media_type=RawOffer";
+
+    private final String RawOfferItemsListSelector = "div.product-list";
+    private final String priceBlockSelector = "div.pl-buy";
+    private final String priceLineSelector = "span.price_lrg.text-cta";
+    private final String RawOfferItemSelector = "div.dv-item";
+    private final String imageLinkSelector = "img.lazy_img.img-fluid";
+    private final String RawOfferInfoBlockSelector = "div.pl-info";
+    private final String RawOfferInfoItemsSelector = "div.vi-text";
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: 1) All variables names from starts from lower case
2) We will have one instance of the class. We can make all variables static constants.

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java (hunk 2 above)
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: Variable names starts from lower case

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java (hunk 7 above)
Comment (AlexKievDeveloper) [reply src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: We should discuss why you need to return null

**Review by AlexKievDeveloper — CHANGES_REQUESTED** [review]

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    Set<String> getPresentPageLinks() {
+        var result = getDocument(startLink);
+        var pages = result
+                .select("a")
+                .stream()
+                .filter(supposedPageLink -> supposedPageLink.text().matches("[0-9]+"))
+                .filter(pageLink -> pageLink.attr("href").startsWith(startBaseLink))
+                .map(pageLink -> pageLink.attr("href"))
+                .collect(toSet());
+        pages = getFullPageLinksList(pages);
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: We shouldn`t reuse variables

**Review by AlexKievDeveloper — CHANGES_REQUESTED** [review]

**Review by Greengrer — DISMISSED** [review]

**Review by Greengrer — DISMISSED** [review]

File: src/test/java/com/vinylteam/vinyl/util/impl/JunoVinylParserTest.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Test
+    void givenLinksToPages_whenRawOffersAreReceived_thenCorrect() {
+        var items = parser.readVinylsDataFromAllPages(Set.of("https://www.juno.co.uk/all/back-cat/3/?media_type=vinyl", "https://www.juno.co.uk/all/back-cat/12/?media_type=vinyl"));
+        assertFalse(items.isEmpty());
+    }
+
+    @Test
+    public void givenHtmlImgElement_whenRawOfferISCreated_thenCorrect() throws IOException {
```
Comment (AlexKievDeveloper) [inline src/test/java/com/vinylteam/vinyl/util/impl/JunoVinylParserTest.java:52]: Why in  the test names you use CamelCase with under lines? Please use only CamelCase. 
Please add @DisplayName annotation on the top of each test to describe what exactly test doing

Comment (TEdwardJ) [reply src/test/java/com/vinylteam/vinyl/util/impl/JunoVinylParserTest.java:52]: just in order to follow some naming strategy for tests methods: https://dzone.com/articles/7-popular-unit-test-naming

@DisplayName - got it, will apply

**Review by AlexKievDeveloper — APPROVED** [review]

**Review by Greengrer — APPROVED** [review]

**Pull request merged**